internal/system/agent/services.go

```diff
@@ -17,165 +17,78 @@ package agent
-					clientInfo := config.ClientInfo{}
-					clientInfo.Protocol = Configuration.Service.Protocol
-					clientInfo.Host = e.Host
-					clientInfo.Port = e.Port
-
-					// This code will evolve to take into account a manifest-like functionality in future. So
-					// rather than assume that the runtime bool flag useRegistry has been initialized to true,
-					// given that the flow has reached this point, having already called functions on the Registry,
-					// such as registryClient.IsServiceAvailable(service), we test for its truthiness. I expect
-					// this code to be refactored as we evolve toward a manifest-like functionality in future.
-					usingRegistry := false
-					if registryClient != nil {
-						usingRegistry = true
-					}
-
-					Configuration.Clients[e.ServiceId] = clientInfo
-					params := types.EndpointParams{
-						ServiceKey:  e.ServiceId,
-						Path:        "/",
-						UseRegistry: usingRegistry,
-						Url:         Configuration.Clients[e.ServiceId].Url() + clients.ApiConfigRoute,
-						Interval:    internal.ClientMonitorDefault,
-					}
-					// TODO: The following note is related to future work:
-					// TODO: With the current deployment strategy, GeneralClient's func init(params types.EndpointParams) {...} returns blank "url"...
-					// TODO: [Need a manifest-like functionality in future...]
-					// Add the service key to the map where the value is the respective GeneralClient
-					generalClients[e.ServiceId] = general.NewGeneralClient(params, startup.Endpoint{RegistryClient: &registryClient})
-
-					responseJSON, err := generalClients[e.ServiceId].FetchConfiguration(ctx)
-					if err != nil {
-						c.Configuration[service] = fmt.Sprintf(err.Error())
-						LoggingClient.Error(err.Error())
-					} else {
-						c.Configuration[service] = ProcessResponse(responseJSON)
-					}
-					return c, nil
-				}
+				result = append(result, executor.Failure(service, metrics, executorTypeDirectService, err.Error()))
+				continue
 			}
-		} else {
-			// Service is known to SMA, so no need to ask the Registry for a ServiceEndpoint associated with `service`
-			// Simply use one of the ready-made list of clients.
-			LoggingClient.Info(fmt.Sprintf("service %s is known to SMA as being in the ready-made list of clients", service))
-
-			responseJSON, err := generalClients[service].FetchConfiguration(ctx)
+			result = append(result, out)
+		case metricsOptionViaExecutor:
+			out, err := MetricsViaExecutor(service)
 			if err != nil {
-				c.Configuration[service] = fmt.Sprintf(err.Error())
-				LoggingClient.Error(err.Error())
-			} else {
-				c.Configuration[service] = ProcessResponse(responseJSON)
+				result = append(result, executor.Failure(service, metrics, executorTypeUnknown, err.Error()))
+				continue
 			}
+			result = append(result, processResponse(out))
+		case metricsOptionViaCustom:
+			return nil, fmt.Errorf("the requested custom executor (e.g. snap) has not been integrated")
```

Comment:
what's this about snap not being integrated? the snap will use the direct service metrics method

Reply:
I understood from Akram there were three distinct metrics gathering implementations; direct-service, executor, and custom.  I ran your comment by him and he concurs that the "custom" option is not required.  I will remove it and push an updated version of the code.

Reply:
yeah AFAIK, there's currently 2 implementations, 1 is a custom binary and 2 is direct service

Reply:
Thanks for the clarification, @anonymouse64 👍

Reply:
Pushed.